### PR TITLE
initial dry run command

### PR DESCRIPTION
**Pull request by DeNeutoy**
This is a WIP and i'm going to add to it incrementally.

Here is what the output currently looks like:
@matt-gardner, @joelgrus, @schmmd, top level comments welcome, other stuff you think needs to be in here from the start.
I'm planning on adding:
- Saved histograms of field lengths
- learning rate graphing functionality
- Model parameter counts
- Keras-like summary of the model
Anything else you think is useful?

```
----Dataset Statistics----

Statistics for premise:
        Lengths: Mean: 12.0, Standard Dev: 0.0
Statistics for hypothesis:
        Lengths: Mean: 10.0, Standard Dev: 0.816496580927726

10 Random instances:
Instance 5:
        Instance with fields:
         premise: TextField of length 12 with text:
                [A, person, on, a, horse, jumps, over, a, broken, down, airplane, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         hypothesis: TextField of length 9 with text:
                [A, person, is, outdoors, ,, on, a, horse, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         label: LabelField with label: entailment in namespace: 'labels'.'

Instance 4:
        Instance with fields:
         premise: TextField of length 12 with text:
                [A, person, on, a, horse, jumps, over, a, broken, down, airplane, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         hypothesis: TextField of length 11 with text:
                [A, person, is, at, a, diner, ,, ordering, an, omelette, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         label: LabelField with label: contradiction in namespace: 'labels'.'

Instance 0:
        Instance with fields:
         premise: TextField of length 12 with text:
                [A, person, on, a, horse, jumps, over, a, broken, down, airplane, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         hypothesis: TextField of length 10 with text:
                [A, person, is, training, his, horse, for, a, competition, .]
                and TokenIndexers : {'tokens': 'SingleIdTokenIndexer'}
         label: LabelField with label: neutral in namespace: 'labels'.'

etc....


----Vocabulary Statistics----

Vocabulary with namespaces:
        Non Padded Namespaces: ('*tags', '*labels')
        Namespace: tokens, Size: 24
        Namespace: labels, Size: 3


Top 10 most frequent tokens in namespace 'tokens':
        Token: a                Frequency: 30
        Token: person           Frequency: 12
        Token: .                Frequency: 12
        Token: horse            Frequency: 10
        Token: on               Frequency: 8
        Token: jumps            Frequency: 6
        Token: over             Frequency: 6
        Token: broken           Frequency: 6
        Token: down             Frequency: 6
        Token: airplane         Frequency: 6

Top 10 longest tokens in namespace 'tokens':
        Token: competition              length: 11      Frequency: 2
        Token: airplane         length: 8       Frequency: 6
        Token: training         length: 8       Frequency: 2
        Token: ordering         length: 8       Frequency: 2
        Token: omelette         length: 8       Frequency: 2
        Token: outdoors         length: 8       Frequency: 2
        Token: person           length: 6       Frequency: 12
        Token: broken           length: 6       Frequency: 6
        Token: horse            length: 5       Frequency: 10
        Token: jumps            length: 5       Frequency: 6

Top 10 shortest tokens in namespace 'tokens':
        Token: ,                length: 1       Frequency: 4
        Token: .                length: 1       Frequency: 12
        Token: a                length: 1       Frequency: 30
        Token: an               length: 2       Frequency: 2
        Token: at               length: 2       Frequency: 2
        Token: is               length: 2       Frequency: 6
        Token: on               length: 2       Frequency: 8
        Token: for              length: 3       Frequency: 2
        Token: his              length: 3       Frequency: 2
        Token: down             length: 4       Frequency: 6

Top 10 most frequent tokens in namespace 'labels':
        Token: neutral          Frequency: 2
        Token: contradiction            Frequency: 2
        Token: entailment               Frequency: 2

Top 10 longest tokens in namespace 'labels':
        Token: contradiction            length: 13      Frequency: 2
        Token: entailment               length: 10      Frequency: 2
        Token: neutral          length: 7       Frequency: 2

Top 10 shortest tokens in namespace 'labels':
        Token: neutral          length: 7       Frequency: 2
        Token: entailment               length: 10      Frequency: 2
        Token: contradiction            length: 13      Frequency: 2
2018-04-03 16:11:42,721 - INFO - allennlp.commands.dry_run - writing the vocabulary to ./stats-test.
2018-04-03 16:11:42,721 - WARNING - root - vocabulary serialization directory ./stats-test/vocabulary is not empty
2018-04-03 16:11:42,721 - INFO - allennlp.commands.dry_run - done creating vocab

```

### CODE REVIEW

**Comment by matt-gardner** [comment]
The "lengths" output seems like it's missing the padding keys.  Also could use a "max" output, not just mean and std.  Other than that, looks great!  Maybe call this `data-stats`, or something, instead of `dry-run`?

File: allennlp/commands/dry_run.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    vocab_params : ``Params``, required
+        The parameters for the Vocabulary we create.
+    instances : ``List[Instance]``, required.
+        The instances to build the vocabulary from.
+
+    Returns
+    -------
+    The created Vocabulary.
+    """
+    min_count = vocab_params.pop("min_count", None)
```
Comment (matt-gardner) [inline allennlp/commands/dry_run.py]: Why can't you just call `Vocabulary.from_params` here?  It's because you want to have access to `namespace_token_counts` when you're done, right?  If that's true, I'd make sure to note it in a comment here, to explain why there's a bunch of duplicated code.

File: allennlp/commands/dry_run.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                sequence_field_lengths[f"{field}.{key}"].append(value)
+
+    print("\n\n----Dataset Statistics----\n")
+    for name, lengths in sequence_field_lengths.items():
+        print(f"Statistics for {name}:")
+        print(f"\tLengths: Mean: {numpy.mean(lengths)}, Standard Dev: {numpy.std(lengths)}, "
+              f"Max: {numpy.max(lengths)}, Min: {numpy.min(lengths)}")
+
+    print("\n10 Random instances: ")
+    for i in list(numpy.random.randint(len(instances), size=10)):
```
Comment (matt-gardner) [inline allennlp/commands/dry_run.py]: Really doesn't matter (i.e., feel free to totally ignore this), but maybe slightly more pythonic:

```python
for i, instance in enumerate(random.sample(instances, 10)):
    ...
```

Oh, except you're outputting the original index, not "first random example", "second random example", etc. ...  Ok, yeah, what I said won't do that.

File: tests/data/vocabulary_test.py [hunk 3]
```diff
@@ -300,3 +300,12 @@ def test_from_params(self):
...
             _ = Vocabulary.from_params(Params({"directory_path": vocab_dir, "min_count": {'tokens': 2}}))
+
+    def test_vocab_can_print(self):
+        vocab = Vocabulary(non_padded_namespaces=["a", "c"])
+        vocab.add_token_to_namespace("a0", namespace="a")
+        vocab.add_token_to_namespace("a1", namespace="a")
+        vocab.add_token_to_namespace("a2", namespace="a")
+        vocab.add_token_to_namespace("b2", namespace="b")
+        vocab.add_token_to_namespace("b3", namespace="b")
+        print(vocab)
```
Comment (matt-gardner) [inline tests/data/vocabulary_test.py]: You're really testing the `__str__` method here, and it'd be more informative to make sure that the string representation has a few key things that you care about.  i.e., instead of just calling `print(vocab)`, do some `contains` asserts on `str(vocab)`.

File: allennlp/commands/dry_run.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
+"""
+The ``dry-run`` command creates a vocabulary, informs you of
```
[the remark this answers is not in the document]
Comment (DeNeutoy) [reply allennlp/commands/dry_run.py]: I guess so? Can't really see why it wouldn't, I'll remove it in a separate PR once this is completely finished.

File: tests/data/vocabulary_test.py (hunk 3 above)
Comment (DeNeutoy) [reply tests/data/vocabulary_test.py]: I'm not really testing it's content, just that it doesn't crash. It's the same as all the other __str__ tests and seems a bit fine grained to test whether a string get's mutated from its string representation in between being called and printing it.

Comment (matt-gardner) [reply tests/data/vocabulary_test.py]: The reason I brought it up was because my initial thought was to say "Remove print statement" and move on.  But then I realized the print was actually part of the test.  But you don't _need_ the `print` for the test, just `str(vocab)`.
